File: Civitas-Java/Civitas/src/main/java/civitas/CivitasJuego.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import GUI.Dado;
import java.util.ArrayList;
import java.util.Collections;

/**
 * @author elena
 * @date 04/10/2019
 */

public class CivitasJuego {
    private int indiceJugador;
    private Tablero tablero;
    private MazoSorpresas mazo;
    private EstadosJuego estado;
    private GestorEstados gestorEstados;
    private ArrayList<Jugador> jugadores;
    private int i = 0;

    public CivitasJuego(ArrayList<String> nombres){
        jugadores = new ArrayList();
        
        for(String n : nombres){
            jugadores.add(new Jugador(n));
        }
        
        gestorEstados = new GestorEstados();
        estado = gestorEstados.estadoInicial();
        
        indiceJugador = Dado.getInstance().quienEmpieza(jugadores.size()-1);
        mazo = new MazoSorpresas();
        this.inicializarTablero(mazo);
        this.inicializarMazoSorpresas(tablero);
    }
    
    private void inicializarTablero(MazoSorpresas mazo){
        tablero = new Tablero(5);
        
        tablero.añadeCasilla(new Casilla("Salida"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Recogidas",
        50, 20, 100, 100, 50)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Tranvia",
        55, 20, 120, 150, 60)));
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa1"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Aeropuerto",
        60, 20, 140, 200, 70)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Sana",
        65, 20, 160, 250, 80)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Ole",
        80, 20, 200, 300, 90)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Mesa",
        85, 20, 220, 350, 100)));
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa2"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Trompeta",
        90, 20, 240, 400, 110)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Maracas",
        95, 20, 260, 450, 120)));
        tablero.añadeCasilla(new CasillaImpuesto(400, "Impuesto"));//impuesto
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Solida",
        110, 20, 300, 500, 130)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Coche",
        115, 20, 320, 550, 140)));
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa3"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Pienso",
        120, 20, 340, 600, 150)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Tren",
        125, 20, 360, 650, 160)));
        tablero.añadeJuez();//juez
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Mal",
        140, 20, 400, 700, 170)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Guitarra",
        145, 20, 420, 750, 180)));
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa4"));
        tablero.añadeCasilla(new Casilla("Parking"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Hora",
        155, 20, 440, 800, 190)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Diamante",
        170, 20, 460, 850, 200)));        
    }
    
    private void inicializarMazoSorpresas(Tablero tablero){
        mazo.alMazo(new SorpresaIRCARCEL(tablero));
        mazo.alMazo(new SorpresaIRCARCEL(tablero));
        mazo.alMazo(new SorpresaIRCASILLA(tablero, 4, "Ve a la casilla 4"));
        mazo.alMazo(new SorpresaIRCASILLA(tablero, 12, "Ve a la casiila 12"));
        mazo.alMazo(new SorpresaPAGARCOBRAR(-500, "Paga a otro jugador"));
        mazo.alMazo(new SorpresaPAGARCOBRAR(500, "Recibes dinero *-*"));
        mazo.alMazo(new SorpresaPORCASAHOTEL(100, "Pagas por cada casa y hotel"));
        mazo.alMazo(new SorpresaPORJUGADOR(50, "Recibes dinero de los demas"));
        mazo.alMazo(new SorpresaSALIRCARCEL(mazo));
        mazo.alMazo(new SorpresaSALIRCARCEL(mazo));
        mazo.alMazo(new SorpresaPORJUGADOR(-50, "Dale dinero a los demás"));
        mazo.alMazo(new SorpresaPAGARCOBRAR(-100, "Pagas a la banca"));                           
    }
    
    private void contabilizarPasosPorSalida(Jugador jugadorActual){
        while(tablero.getPorSalida() > 0){
            jugadorActual.pasaPorSalida();
        }
    }
    
    private void pasarTurno(){ indiceJugador = (indiceJugador+1)%jugadores.size(); }
    
    public void siguientePasoCompletado(OperacionesJuego operacion){
        this.estado = gestorEstados.siguienteEstado(jugadores.get(indiceJugador), estado, operacion);
    }
    
    public boolean construirCasa(int ip){ return jugadores.get(indiceJugador).construirCasa(ip); }
    
    public boolean construirHotel(int ip){ return jugadores.get(indiceJugador).construirHotel(ip); }
    
    public boolean vender(int ip){ return jugadores.get(indiceJugador).vender(ip); }
      
    public boolean hipotecar(int ip){ return jugadores.get(indiceJugador).hipotecar(ip); }
    
    public boolean cancelarHipoteca(int ip){ return jugadores.get(indiceJugador).cancelarHipoteca(ip); }
    
    public boolean salirCarcelPagando(){ return jugadores.get(indiceJugador).salirCarcelPagando(); }
    
    public boolean salirCarcelTirando(){ return jugadores.get(indiceJugador).salirCarcelTirando(); }
    
    public boolean finalDelJuego(){ 
        for(Jugador j : jugadores){
            if(j.enBancarrota()) return true;
        }
        return false;               
    }
    
    //Visibilidad cambiada
    public ArrayList<Jugador> ranking(){ 
        Collections.sort(jugadores);
        return jugadores;
    }
    
    public Casilla getCasillaActual(){ return tablero.getCasilla(getJugadorActual().getNumCasillaActual()); }
    
    public Jugador getJugadorActual(){ return jugadores.get(indiceJugador); }    
    
    public String infoJugadorTexto(){ return jugadores.get(indiceJugador).toString(); }   
        
    private void avanzaJugador(){
        Jugador jugadorActual = jugadores.get(indiceJugador);   //1.1
        
        int posicionActual = jugadorActual.getNumCasillaActual();  //1.2
        int tirada = Dado.getInstance().tirar();    //1.3
        int posicionNueva = tablero.nuevaPosicion(posicionActual, tirada);  //1.4
        Casilla casilla = tablero.getCasilla(posicionNueva);    //1.5
        this.contabilizarPasosPorSalida(jugadorActual); //1.6
        jugadorActual.moverACasilla(posicionNueva); //1.7
        casilla.recibeJugador(indiceJugador, jugadores);    //1.8
        this.contabilizarPasosPorSalida(jugadorActual); //1.9
    }      
    
    public boolean comprar(){ 
        boolean res = false;
        Jugador jugadorActual = jugadores.get(indiceJugador);
        int numCasillaActual = jugadorActual.getNumCasillaActual();
        
        if(tablero.getCasilla(numCasillaActual).soyCasillaCalle()){
            CasillaCalle casilla = (CasillaCalle) tablero.getCasilla(numCasillaActual);
            TituloPropiedad titulo = casilla.getTituloPropiedad();
            res = jugadorActual.comprar(titulo);
        }               
        return res;
    }   
    
    public OperacionesJuego siguientePaso(){ 
        Jugador jugadorActual = jugadores.get(indiceJugador);
        OperacionesJuego operacion = gestorEstados.operacionesPermitidas(jugadorActual, estado);
        
        if(operacion == OperacionesJuego.PASAR_TURNO){
            pasarTurno();
            siguientePasoCompletado(operacion);
        }
        else if(operacion == OperacionesJuego.AVANZAR){
            avanzaJugador();
            siguientePasoCompletado(operacion);
        }
        
        return operacion;
    }
}

File: Civitas-Ruby/lib/civitas_juego.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "tablero"
require_relative "operaciones_juego"
require_relative "mazo_sorpresas"
require_relative "dado"
require_relative "gestor_estados"
require_relative "estados_juego"
require_relative "jugador"

module Civitas
  class CivitasJuego
    def initialize(nombres)
      @jugadores = Array.new
      for n in nombres
        @jugadores << Jugador.new(n)
      end
      
      @gestorestados = Gestor_estados.new
      @estado =  @gestorestados.estado_inicial
      
      @indiceJugadorActual = Dado.instance.tirar
      
      @mazo = MazoSorpresas.new
      inicializartablero(@mazo)
      inicializarmazosorpresas(@tablero)
    end
    
    def cancelarhipoteca(ip)
      getjugadoractual.cancelarhipoteca(ip)
    end
    
    def comprar      
      res = false
      jugador_actual = getjugadoractual
      num_casilla_actual = jugador_actual.numCasillaActual
      casilla = @tablero.getcasilla(num_casilla_actual)
      if casilla.nombre != 'Salida'
        titulo = casilla.titulo
        res = jugador_actual.comprar(titulo)
      end
      res
    end
    
    def construircasa(ip)
      getjugadoractual.construircasa(ip)
    end
    
    def construirhotel(ip)
      getjugadoractual.construirhotel(ip)
    end
    
    def finaldeljuego
      for j in @jugadores
        if j.enbancarrota
          true
        end
      end
      false
    end
    
    def getcasillaactual      
      @tablero.getcasilla(getjugadoractual.numCasillaActual)
    end
    
    def getjugadoractual
      @jugadores.at(@indiceJugadorActual)
    end
    
    def hipotecar(ip)
      getjugadoractual.hipotecar(ip)
    end
    
    def infojugadortexto
      getjugadoractual.to_s
    end
    
    def salircarcelpagando
      getjugadoractual.salircarcelpagando(ip)
    end
    
    def salircarceltirando
      getjugadoractual.salircarceltirando(ip)
    end
    
    def siguientepaso    
      jugador_actual = getjugadoractual
      operacion = @gestorestados.operaciones_permitidas(jugador_actual, @estado)
      
      if operacion == Civitas::Operaciones_juego::PASAR_TURNO
        pasarturno
        siguientepasocompletado(operacion)
      else if operacion == Civitas::Operaciones_juego::AVANZAR
        avanzajugador
        siguientepasocompletado(operacion)
      end
      end
      
      operacion
    end
    
    def siguientepasocompletado(operacion)
      @estado = @gestorestados.siguiente_estado(getjugadoractual, @estado, operacion)
    end
    
    def vender(ip)
      getjugadoractual.vender(ip)
    end
    
    private
    def avanzajugador
      jugador_actual = getjugadoractual  #1.1
      posicion_actual = jugador_actual.numCasillaActual #1.2
      tirada = Dado.instance.tirar  #1.3
      posicion_nueva = @tablero.nuevaposicion(posicion_actual, tirada)  #1.4
      casilla = @tablero.getcasilla(posicion_nueva)   #1.5
      contabilizarpasosporsalida(jugador_actual)  #1.6
      jugador_actual.moveracasilla(posicion_nueva) #1.7
      casilla.recibejugador(@indiceJugadorActual, @jugadores) #1.8
      contabilizarpasosporsalida(jugador_actual)  #1.9
    end
    
    def contabilizarpasosporsalida(jugadoractual)
      if @tablero.porSalida > 0
        jugadoractual.pasaporsalida
      end
    end
    
    def inicializarmazosorpresas(tablero)
      @mazo.almazo(Sorpresa.new_sorpresa_ircarcel(Civitas::TipoSorpresa::IRCARCEL, tablero))
      @mazo.almazo(Sorpresa.new_sorpresa_ircarcel(Civitas::TipoSorpresa::IRCARCEL, tablero))
      @mazo.almazo(Sorpresa.new_sorpresa_casilla(Civitas::TipoSorpresa::IRCASILLA, tablero, 4, 'Ve a la casilla 4'))
      @mazo.almazo(Sorpresa.new_sorpresa_casilla(Civitas::TipoSorpresa::IRCASILLA, tablero, 12, 'Ve a la casilla 12'))
      @mazo.almazo(Sorpresa.new_sorpresas(Civitas::TipoSorpresa::PAGARCOBRAR, -500, 'Paga a otro jugador'))
      @mazo.almazo(Sorpresa.new_sorpresas(Civitas::TipoSorpresa::PAGARCOBRAR, 500, 'Recibes dinero *-*'))
      @mazo.almazo(Sorpresa.new_sorpresas(Civitas::TipoSorpresa::PORCASAHOTEL, 100, 'Pagas por cada casa y hotel'))
      @mazo.almazo(Sorpresa.new_sorpresas(Civitas::TipoSorpresa::PORJUGADOR, 50, 'Recibes dinero de los demas'))
      @mazo.almazo(Sorpresa.new_sorpresas(Civitas::TipoSorpresa::PORJUGADOR, -50, 'Dale dinero a los demas'))
      @mazo.almazo(Sorpresa.new_sorpresa_evitarcarcel(Civitas::TipoSorpresa::SALIRCARCEL, @mazo))
      @mazo.almazo(Sorpresa.new_sorpresa_evitarcarcel(Civitas::TipoSorpresa::SALIRCARCEL, @mazo))
      @mazo.almazo(Sorpresa.new_sorpresa_evitarcarcel(Civitas::TipoSorpresa::SALIRCARCEL, @mazo))                          
    end
    
    def inicializartablero(mazo)
      @tablero = Tablero.new(5)
      
      @tablero.aniadecasilla(Casilla.new_casilla('Salida'))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Recogidas',
      50, 20, 100, 100, 50)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Tranvia',
      55, 20, 120, 150, 60)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Aeropuerto',
      60, 20, 140, 200, 70)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Sana',
      65, 20, 160, 250, 80)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Ole',
      80, 20, 200, 300, 90)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Mesa',
      85, 20, 220, 350, 100)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Trompeta',
      90, 20, 240, 400, 110)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Maracas',
      95, 20, 260, 450, 120)))
      @tablero.aniadecasilla(Casilla.new_casillaimpuesto(400, 'Impuesto'))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Solida',
      110, 20, 300, 500, 130)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Coche',
      115, 20, 320, 550, 140)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Pienso',
      120, 20, 340, 600, 150)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Tren',
      125, 20, 360, 650, 160)))
      @tablero.aniadejuez
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Mal',
      140, 20, 400, 700, 170)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Guitarra',
      145, 20, 420, 750, 180)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Hora',
      155, 20, 440, 800, 190)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Diamante',
      170, 20, 460, 850, 200))) 
    end
    
    def pasarturno
      @indiceJugadorActual = (@indiceJugadorActual+1)%@jugadores.length
    end
    
    public
    def ranking
      @jugadores.sort
    end
  end
end
